packages/doenetml-worker-rust/lib-doenetml-core/src/core/components/special/_external.rs

```diff
@@ -123,4 +128,34 @@ impl ComponentNode for _External {
     fn get_component_profile_state_variables(&self) -> &Vec<ComponentProfileStateVariable> {
         &self.common.component_profile_state_variables
     }
+
+    fn set_attribute_children(
+        &mut self,
+        attribute_children: HashMap<AttributeName, Vec<ComponentPointerTextOrMacro>>,
+    ) {
+        self.common.attribute_children = attribute_children;
+    }
+
+    fn get_attribute_children_for_attribute(
+        &self,
+        attribute: AttributeName,
+    ) -> Option<&Vec<ComponentPointerTextOrMacro>> {
+        self.common.attribute_children.get(attribute)
+    }
+
+    fn get_unevaluated_attributes(&self) -> &HashMap<String, DastAttribute> {
+        &self.common.unevaluated_attributes
+    }
+
+    fn get_unevaluated_attributes_mut(&mut self) -> &mut HashMap<String, DastAttribute> {
+        &mut self.common.unevaluated_attributes
+    }
+
+    fn get_is_rendered(&self) -> bool {
```

Comment:
I'm still confused by `rendered`. I forget what it means, but `rendered` could mean "will be rendered" or could mean "has been rendered"

Reply:
It means neither. It specifies a fixed attribute of a component rather than a state. A `rendered` component is a component that will/does appear in the flat dast tree. A `non-rendered` component still appears in the `flat_dast` because the `flat_dast` contains all components even if they aren't in the tree. But, we won't spend the cpu cycles calculating the rendered state for a component it if is not `rendered`.

For state variables, I used the attribute `for_renderer` to distinguish which variables are in the rendered data. Maybe for components, as in here, I'll call it `in_render_tree`. Would that be clearer?